Chap_Introduction.tex

```diff
@@ -232,4 +234,81 @@ \subsection{PMIx Roles}
 These \emph{client-role,server-role, and tool-role \ac{PMIx} standard compliant} implementations have the advantage of being able to support a broad set of \ac{PMIx} consumers in the different roles.
 \adviceimplend
 
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+\subsection{Application Binary Interface (ABI)}
+\label{chap:intro:abi}
+
+An \acf{API} defines how data types, data structures, and functions are represented in source code.
+An \acf{ABI} defines how data types, data structures, and functions are represented in machine code for a given system making it platform-specific.
+An important aspect of an \ac{ABI} is the size, layout, and alignment of data structures.
+A stable \ac{ABI} may allow a program compiled with one implementation of the PMIx Standard to run with a different implementation of the PMIx Standard as long as both implementations adhere to the same \ac{ABI}.
+
+The PMIx Standard strives to maintain a stable \ac{ABI} to support applications and tools that rely on more than one implementation of the PMIx Standard.
+To facilitate such interoperability the PMIx \ac{ASC} maintains a set of standardized headers that are versioned with the PMIx Standard that applications and tools can reference~\footnote{PMIx Headers for ABI Compatibility \url{https://github.com/pmix/pmix-abi}}.
+
+In recognition that there are circumstances where the \ac{ABI} needs to be modified this section defines some guidance for making such modifications.
+Additions to the PMIx interface can occur without breaking \ac{ABI} compatibility.
+Deprecating and removing portions of the PMIx interface does break \ac{ABI} compatibility.
+Modifications to the existing PMIx interface do break \ac{ABI} compatibility.
+Any changes to the \ac{ABI} will occur in a major release and will be noted in the release notes.
+
+The PMIx \ac{ABI} is comprised of the following:
+\begin{compactitemize}
+\item Function signatures
+\item Constants and their values
+\item Structures and their membership (including field position)
+\item Attributes that are required to be implemented and their string representation
+\end{compactitemize}
+
+If a data structure must be changed in a way that breaks the \ac{ABI} the authors should consider creating a new data structure if possible.
```

Comment:
Changing the ABI of data structures (e.g., removing/reordering existing fields) will break backward compatibility. We should add a short rationale as to why this is **very** undesirable, and not only one should *consider*, but should avoid this type of changes at all cost.

Reply:
```suggestion
If a data structure must be changed in a way that breaks the \ac{ABI} the authors are strongly advised to create a new data structure instead of breaking the \ac{ABI} and backward compatibility.
```

Let's add a clear rationale to add to this section to capture the middleware constraints dealing with cross and backwards compatibility issues.

Reply:
```diff
+Changing data structures (in particular changes that modify the field alignment) and function signatures
+breaks the \ac{ABI}. The \ac{ASC} strongly discourages such changes. It is preferred that the authors 
+duplicate existing PMIx interfaces that are impacted by the change, 
+and potentially initiate the deprecation procedure for the duplicated interfaces.
+
+\rationalestart
+Breaking the \ac{ABI} of data structures and function signatures while using the same
+name prevents the implementation of backward compatible and cross-version compatible
+\ac{PMIx} libraries. 
+\rationaleend
```

Reply:
I like the new text but we may need some more plenary time to discuss the implications.

Reply:
```suggestion
The \ac{ASC} prohibits data structure and function signature changes that break the \ac{ABI}
(e.g., modifications to data structure member alignment and/or size).
It is preferred that the authors create a new data structure or function and
duplicate all impacted PMIx interfaces.

\rationalestart
Breaking the \ac{ABI} of data structures and function signatures while using the same
name prevents a PMIx implementation from supporting backward compatible and
cross-version compatible \ac{PMIx} libraries. 
\rationaleend
```